src/app_conn_pool.rs

```diff
@@ -0,0 +1,319 @@
+use crate::config::{AllowedFns, Configuration};
+use crate::{HcHttpGatewayError, HcHttpGatewayResult};
+use futures::future::BoxFuture;
+use holochain_client::{
+    AdminWebsocket, AppWebsocket, AuthorizeSigningCredentialsPayload, CellInfo, ClientAgentSigner,
+    ConductorApiError, ConnectRequest, GrantedFunctions, IssueAppAuthenticationTokenPayload,
+    Timestamp, WebsocketConfig,
+};
+use holochain_types::app::InstalledAppId;
+use holochain_types::websocket::AllowedOrigins;
+use std::collections::HashMap;
+use std::net::SocketAddr;
+use std::sync::{Arc, RwLock};
+
+/// The origin that the gateway will use when connecting to Holochain app interfaces.
+pub const HTTP_GW_ORIGIN: &str = "hc-http-gw";
+
+/// A wrapper around an app websocket connection that includes state required to manage the
+/// connection.
+#[derive(Debug, Clone)]
+pub struct AppWebsocketWithState {
+    /// The app websocket connection.
+    pub app_ws: AppWebsocket,
+    /// The time at which the connection was opened.
+    pub opened_at: Timestamp,
+}
+
+/// A connection pool for app connections.
+///
+/// This is a pool in the sense that it manages multiple connections to Holochain app interfaces,
+/// but it will manage exactly one connection per installed app.
+#[derive(Debug, Clone)]
+pub struct AppConnPool {
+    configuration: Configuration,
+    cached_app_port: Arc<RwLock<Option<u16>>>,
+    app_clients: Arc<tokio::sync::RwLock<HashMap<InstalledAppId, AppWebsocketWithState>>>,
+}
+
+impl AppConnPool {
+    /// Create a new app connection pool with the given configuration.
+    pub fn new(configuration: Configuration) -> Self {
+        Self {
+            configuration,
+            cached_app_port: Default::default(),
+            app_clients: Default::default(),
+        }
+    }
+
+    /// Call a function with an app client for the given installed app ID.
+    ///
+    /// This function takes care of reconnecting to the app client if the connection is lost. Your
+    /// function is free to operate on the app client without worrying about the connection state.
+    ///
+    /// Exactly one attempt is made to reconnect and re-run the provided callback. If there is a
+    /// second failure after reconnecting, the error is returned to the caller.
+    pub async fn call<T>(
+        &self,
+        installed_app_id: InstalledAppId,
+        admin_websocket: AdminWebsocket,
+        execute: impl Fn(AppWebsocket) -> BoxFuture<'static, HcHttpGatewayResult<T>>,
+    ) -> HcHttpGatewayResult<T> {
+        for _ in 0..2 {
+            let app_ws = self
+                .get_or_connect_app_client(installed_app_id.clone(), admin_websocket.clone())
+                .await?;
+            match execute(app_ws).await {
+                Ok(response) => {
+                    return Ok(response);
+                }
+                Err(HcHttpGatewayError::HolochainError(ConductorApiError::WebsocketError(e))) => {
```

Comment:
Cheaper to store the app auth token expiration and check it ourselves before making the websocket call? Error due to token expiration will happen frequently.

Reply:
We're not caching the token and the token lifetime is only used by Holochain to check that it is still valid when the connection is made, there's no need to reconnect later with a new token if the websocket stays open. Because we're using single-use tokens, I don't think it should really happen that we fail with a token error. It's more likely that Holochain closed the connection on a conductor restart or something like that. We might fail to connect if we used a cached port, and hopefully the updated code makes that clearer. Otherwise if we get connection error without anything cached, there's no reason to retry connecting.

https://github.com/holochain/hc-http-gw/pull/17/commits/394963ae93978fd098e03ddf63faefbe7da4d050